src/Avalonia.Base/ValueStore.cs

```diff
@@ -178,5 +195,110 @@ public DeferredSetter<object> Setter
+            {
+                if (_entries[i].PropertyId > property.Id)
+                {
+                    if (i > 0)
+                    {
+                        Array.Copy(_entries, 0, entries, 0, i);
+                    }
+
+                    entries[i] = new Entry { PropertyId = property.Id, Value = value };
+                    Array.Copy(_entries, i, entries, i + 1, _entries.Length - i);
+                    break;
+                }
+            }
+
+            _entries = entries;
+        }
+
+        private void SetValueInternal(AvaloniaProperty property, object value)
+        {
+            _entries[TryFindEntry(property.Id).Item1].Value = value;
+        }
+
+        private (int, bool) TryFindEntry(int propertyId)
+        {
+            if (_entries.Length <= 16)
+            {
+                // For small lists, we use an optimized linear search. Since the last item in the list
+                // is always int.MaxValue, we can skip a conditional branch in each iteration.
+                // By unrolling the loop, we can skip another unconditional branch in each iteration.
+
+                if (_entries[0].PropertyId >= propertyId) return (0, _entries[0].PropertyId == propertyId);
+                if (_entries[1].PropertyId >= propertyId) return (1, _entries[1].PropertyId == propertyId);
+                if (_entries[2].PropertyId >= propertyId) return (2, _entries[2].PropertyId == propertyId);
+                if (_entries[3].PropertyId >= propertyId) return (3, _entries[3].PropertyId == propertyId);
+                if (_entries[4].PropertyId >= propertyId) return (4, _entries[4].PropertyId == propertyId);
+                if (_entries[5].PropertyId >= propertyId) return (5, _entries[5].PropertyId == propertyId);
+                if (_entries[6].PropertyId >= propertyId) return (6, _entries[6].PropertyId == propertyId);
+                if (_entries[7].PropertyId >= propertyId) return (7, _entries[7].PropertyId == propertyId);
+                if (_entries[8].PropertyId >= propertyId) return (8, _entries[8].PropertyId == propertyId);
+                if (_entries[9].PropertyId >= propertyId) return (9, _entries[9].PropertyId == propertyId);
+                if (_entries[10].PropertyId >= propertyId) return (10, _entries[10].PropertyId == propertyId);
+                if (_entries[11].PropertyId >= propertyId) return (11, _entries[11].PropertyId == propertyId);
+                if (_entries[12].PropertyId >= propertyId) return (12, _entries[12].PropertyId == propertyId);
+                if (_entries[13].PropertyId >= propertyId) return (13, _entries[13].PropertyId == propertyId);
+                if (_entries[14].PropertyId >= propertyId) return (14, _entries[14].PropertyId == propertyId);
+            }
+            else
+            {
+                int low = 0;
+                int high = _entries.Length;
+                int id;
+
+                if (high > 0)
```

Comment:
Looks like a binary search. It might be better to utilize `Array.BinarySearch` since it's a well-tested implementation

Reply:
Or at least copy-paste it from her: https://github.com/dotnet/corefx/blob/6a219bfd41bf6b6c411c6e1be05ba66dd876328b/src/Common/src/CoreLib/System/SpanHelpers.BinarySearch.cs#L33 since it's shorter

Reply:
In my tests, Array.BinarySearch is on average 11% slower, while SpanHelpers.BinarySearch is on average 13% slower. My guess is that the slight performance drop is due to both Array.BinarySearch and SpanHelpers.BinarySearch using IComparable to compare the individual items, which may incur additional method calls.

My proposed implementation is slightly more specialized in that it doesn't emit any more code than absolutely necessary.